packages/nodes-base/utils/connection-pool-manager.ts

```diff
@@ -0,0 +1,135 @@
+import { createHash } from 'crypto';
+
+let instance: ConnectionPoolManager;
+
+// 5 minutes
+const ttl = 5 * 60 * 1000;
+
+// 1 minute
+const cleanUpInterval = 60 * 1000;
+
+type Options = {
+	credentials: unknown;
+	nodeType: string;
+	nodeVersion?: string;
+};
+
+export class ConnectionPoolManager {
+	/**
+	 * Gets the singleton instance of the ConnectionPoolManager.
+	 * Creates a new instance if one doesn't exist.
+	 */
+	static getInstance(): ConnectionPoolManager {
+		if (instance) {
+			return instance;
+		}
+
+		instance = new ConnectionPoolManager();
+
+		return instance;
+	}
+
+	/**
+	 * Private constructor that initializes the connection pool manager.
+	 * Sets up cleanup handlers for process exit and stale connections.
+	 */
+	private constructor(
+		private map: Map<
+			string,
+			{ pool: unknown; cleanUpHandler: (pool: unknown) => Promise<void>; lastUsed: number }
+		> = new Map(),
+	) {
+		// Close all SSH connections when the process exits
+		process.on('exit', () => this.onShutdown());
+
+		// Regularly close stale SSH connections
+		setInterval(() => this.cleanupStaleConnections(), cleanUpInterval);
+	}
+
+	/**
+	 * Generates a unique key for connection pool identification.
+	 * Hashes the credentials and node information for security.
+	 */
+	private makeKey({ credentials, nodeType, nodeVersion }: Options): string {
+		// The credential contains decrypted secrets, that's why we hash it.
+		return createHash('sha1')
+			.update(
+				JSON.stringify({
+					credentials,
+					nodeType,
+					nodeVersion,
+				}),
+			)
+			.digest('base64');
+	}
+
+	/**
+	 * Gets or creates a connection pool for the given options.
+	 * Updates the last used timestamp for existing connections.
+	 */
+	async getConnection<T>(
+		options: Options,
+		fallBackHandler: () => Promise<T>,
+		cleanUpHandler: (pool: T) => Promise<void>,
```

Comment:
should we create a new type `GetConnectionOptions` that extends `Options`, but had `fallBackHandler` and `cleanUpHandler` as fields?
That would make the call to `getConnection` in the nodes code a lot more readable.